src/mlpack/core/cereal/pointer_variant_wrapper.hpp

```diff
@@ -0,0 +1,114 @@
+/**
+ * @file core/cereal/pointer_variant_wrapper.hpp
+ * @author Omar Shrit
+ *
+ * Implementation of a boost::variant wrapper to enable the serialization of
+ * the pointers inside boost variant in cereal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+#define MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+
+#include <cereal/archives/json.hpp>
+#include <cereal/archives/portable_binary.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/types/boost_variant.hpp>
+
+#include <boost/variant.hpp>
+#include <boost/variant/variant_fwd.hpp>
+#include <boost/variant/static_visitor.hpp>
+
+#include "pointer_wrapper.hpp"
+
+namespace cereal {
+
+template<class Archive>
+struct save_visitor : public boost::static_visitor<void>
+{
+  save_visitor(Archive& ar) : ar_(ar) {}
+  
+  template<class T>
+  void operator()(const T* value) const 
+  {
+    ar_ & CEREAL_POINTER(value);
+  }
+
+  Archive& ar_;
+};
+
+struct load_visitor : public boost::static_visitor<void>
+{
+  template<class Archive, class VariantType>
+  void operator()(Archive& ar, VariantType* variant) const 
```

Comment:
I think that we could maybe use simple overloading here to add an overload of `operator()` for when a `boost::variant` is held internally:

```
template<typename Archive, typename... VariantTypes>
void operator()(Archive& ar, boost::variant<VariantTypes...>& variant) const
{
  ar(CEREAL_VARIANT_POINTER(variant));
}
```

I *think* that would work (and you could do a similar thing for `save_visitor`).  Let me know what you think---I'm happy to try and help debug it if it doesn't work.  The basic idea, I think, should be valid though: we can just catch when a `variant` holds a `variant` inside of itself, and then call `CEREAL_VARIANT_POINTER()` again.

Reply:
@rcurtin I have tested it, but I got the exact same error above. 
I am not sure we can call another `operator()` if the first one has already been called.
I think the idea is valid if there was only `boost::variant<boost::variant<T...>>`. However, since we are mixing type, I am not sure.

Reply:
I am looking through the code now.  Why haven't you used `LoadAndConstructLoadWrapper` like in the `cereal/types/boost_variant.hpp` implementation?  Maybe there is something I have overlooked?  There aren't very many comments, so I'm not sure what the state of the code is---does it fully work and pass tests when serializing a `boost::variant` that doesn't internally contain a `boost::variant`?

Reply:
@rcurtin I think I did not get the reason why they have to use it in the first place.
I think the implementation of `boost::variant` is working since the only errors I am getting now are related to non serialized raw pointers.

Reply:
Okay, but when you pass a `boost::variant<>` that contains pointer types and *also* another `boost::variant`, this still fails?

Reply:
Yes it still fails, actually, it is not complaining about the pointers because if it is the case the compiler would have stopped before. The issue is it is not matching the function call when it sees `boost::variant`
```
double> >*, mlpack::ann::VRClassReward<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::VirtualBatchNorm<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::RBF<arma::Mat<double>, arm
a::Mat<double>, mlpack::ann::GaussianFunction>*, mlpack::ann::BaseLayer<mlpack::ann::GaussianFunction, arma::Mat<double>, arma::Mat<double> >*> > >::value_type&)’
  122 | #define CEREAL_VARIANT_POINTER(T) cereal::make_pointer_variant(T)
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
/meta/mlpack/src/mlpack/methods/ann/ffn_impl.hpp:568:9: note: in expansion of macro ‘CEREAL_VARIANT_POINTER’
  568 |    ar & CEREAL_VARIANT_POINTER(network.at(i));
      |         ^~~~~~~~~~~~~~~~~~~~~~
/meta/mlpack/src/mlpack/core/cereal/pointer_variant_wrapper.hpp:117:1: note: candidate: ‘template<class ... VariantTypes> cereal::pointer_variant_wrapper<VariantTypes ...> cereal::make_point
er_variant(boost::variant<VariantTypes* ...>&)’
  117 | make_pointer_variant(boost::variant<VariantTypes*...>& t)
      | ^~~~~~~~~~~~~~~~~~~~
/meta/mlpack/src/mlpack/core/cereal/pointer_variant_wrapper.hpp:117:1: note:   template argument deduction/substitution failed:
/meta/mlpack/src/mlpack/core/cereal/pointer_variant_wrapper.hpp:122:63: note:   mismatched types ‘VariantTypes*’ and ‘boost::variant<mlpack::ann::Glimpse<arma::Mat<double>, arma::Mat<double>
 >*, mlpack::ann::Highway<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Recurrent<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::RecurrentAttention<arma::Mat<double>, arma::Ma
t<double> >*, mlpack::ann::ReinforceNormal<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Reparametrization<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Select<arma::Mat<doub
le>, arma::Mat<double> >*, mlpack::ann::Sequential<arma::Mat<double>, arma::Mat<double>, false>*, mlpack::ann::Sequential<arma::Mat<double>, arma::Mat<double>, true>*, mlpack::ann::Subview<a
rma::Mat<double>, arma::Mat<double> >*, mlpack::ann::VRClassReward<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::VirtualBatchNorm<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann
::RBF<arma::Mat<double>, arma::Mat<double>, mlpack::ann::GaussianFunction>*, mlpack::ann::BaseLayer<mlpack::ann::GaussianFunction, arma::Mat<double>, arma::Mat<double> >*>’
  122 | #define CEREAL_VARIANT_POINTER(T) cereal::make_pointer_variant(T)
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
/meta/mlpack/src/mlpack/methods/ann/ffn_impl.hpp:568:9: note: in expansion of macro ‘CEREAL_VARIANT_POINTER’
  568 |    ar & CEREAL_VARIANT_POINTER(network.at(i));
      |         ^~~~~~~~~~~~~~~~~~~~~~
[ 80%] Building CXX object src/mlpack/tests/CMakeFiles/mlpack_test.dir/main_tests/decision_stump_test.cpp.o
```

Reply:
Ok, I'll play with this and see if I can get something figured out, by adapting the idea above. :+1:

Reply:
I believe that this implementation works:

```
/**
 * @file core/cereal/pointer_variant_wrapper.hpp
 * @author Omar Shrit
 *
 * Implementation of a boost::variant wrapper to enable the serialization of
 * the pointers inside boost variant in cereal
 *
 * mlpack is free software; you may redistribute it and/or modify it under the
 * terms of the 3-clause BSD license.  You should have received a copy of the
 * 3-clause BSD license along with mlpack.  If not, see
 * http://www.opensource.org/licenses/BSD-3-Clause for more information.
 */
#ifndef MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
#define MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP

#include <cereal/archives/json.hpp>
#include <cereal/archives/portable_binary.hpp>
#include <cereal/archives/xml.hpp>
#include <cereal/types/boost_variant.hpp>

#include <boost/variant.hpp>
#include <boost/variant/variant_fwd.hpp>
#include <boost/variant/static_visitor.hpp>

#include "pointer_wrapper.hpp"

namespace cereal {

// Forward declaration.
template<typename... VariantTypes>
class pointer_variant_wrapper;

template<typename... VariantTypes>
inline pointer_variant_wrapper<VariantTypes...>
make_pointer_variant(boost::variant<VariantTypes...>& t)
{
  return pointer_variant_wrapper<VariantTypes...>(t);
}

template<class Archive>
struct save_visitor : public boost::static_visitor<void>
{
  save_visitor(Archive& ar) : ar_(ar) {}

  template<class T>
  void operator()(const T* value) const~
  {
    ar_ & CEREAL_POINTER(value);
  }

  template<typename... Types>
  void operator()(boost::variant<Types*...>& value) const
  {
    ar_ & make_pointer_variant(value);
  }

  Archive& ar_;
};

template<typename T>
struct load_visitor : public boost::static_visitor<void>
{
  template<typename Archive, typename VariantType>
  static void load_impl(Archive& ar, VariantType& variant, std::true_type)
  {
    // Note that T will be a pointer type.
    T loadVariant;
    ar & CEREAL_POINTER(loadVariant);
    variant = loadVariant;
  }

  template<typename Archive, typename VariantType>
  static void load_impl(Archive& ar, VariantType& value, std::false_type)
  {
    // This must be a nested boost::variant.
    ar & make_pointer_variant(value);
  }

  template<typename Archive, typename VariantType>
  static void load(Archive& ar, VariantType& variant)
  {
    // Delegate to the proper load_impl() overload depending on whether T is a
    // pointer type.  If T is not a pointer type, then we expect it to be a
    // nested boost::variant.
    load_impl(ar, variant, typename std::is_pointer<T>::type() );
  }
};

template<typename... VariantTypes>
class pointer_variant_wrapper
{
/*
 * The objective of this class is to create a wrapper for
 * boost::variant.~
 * Cereal supports the serialization of boost::vairnat, but~
 * we need to serialize it if it holds a raw pointers.
 */
 public:
   pointer_variant_wrapper(boost::variant<VariantTypes...>& PointerVar)
    : PointerVariant(PointerVar)
  {}

  template<class Archive>
  void save(Archive& ar) const
  {
    // which represent the index in std::variant.
    int which = PointerVariant.which();
    ar & CEREAL_NVP(which);
    save_visitor<Archive> s(ar);
    boost::apply_visitor(s, PointerVariant);
  }

  template<class Archive>
  void load(Archive& ar)
  {
    // Load the size of the serialized type.
    int which;
    ar & CEREAL_NVP(which);

    // Create function pointers to each overload of load_visitor<T>::load, for
    // all T in VariantTypes.
    using LoadFuncType = void(*)(Archive&, boost::variant<VariantTypes...>&);
    LoadFuncType loadFuncArray[] = { &load_visitor<VariantTypes>::load... };

    if(which >= int(sizeof(loadFuncArray)/sizeof(loadFuncArray[0])))
      throw std::runtime_error("Invalid 'which' selector when deserializing boost::variant");

    loadFuncArray[which](ar, PointerVariant);
  }

private:
   boost::variant<VariantTypes...>& PointerVariant;
};

#define CEREAL_VARIANT_POINTER(T) cereal::make_pointer_variant(T)

} // end namespace cereal

#endif // CEREAL_POINTER_VARIANT_WRAPPER_HPP
```

However, I did not compile successfully, as there are layers in `src/mlpack/methods/ann/layer/` that internally hold `LayerTypes` or `std::vector<LayerTypes>`, and the code there will need to be adapted to appropriately use the variant wrapper here.  You can find the necessary classes just by doing a `grep LayerTypes` in `src/mlpack/methods/ann/layer/` and you'll see that there are probably 8 or 9 classes that would need to be adapted.  I didn't try, since I wasn't sure if you had a nice wrapper for `std::vector<boost::variant<...>>` or anything.  Maybe another one needs to be made for convenience, like `CEREAL_VECTOR_VARIANT_POINTER()` or something?

Reply:
(I don't mind adding lots of those wrappers.  Eventually, we'll be replacing and removing them anyway, most likely.)

Reply:
I will give it a try and see what will happen.

Reply:
By the way, `CEREAL_VECTOR_VARIANT_POINTER()` will be a true wrapper :+1: :+1: :+1: